File: app/src/main/java/xyz/okxy/hanoitower/DiskView.java
package xyz.okxy.hanoitower;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;

import java.util.Random;

/**
 * 用于动态生成的圆盘View
 * @author zcp
 * @since 2018/9/7
 */
public class DiskView extends View {

    private Context mContext;
    private Paint mPaint;
    private int mIndex;

    public DiskView(Context context, int index) {
        super(context);
        mContext = context;
        mPaint = new Paint();
        Random random = new Random();
        int color = Color.rgb(
                random.nextInt(256),
                random.nextInt(256),
                random.nextInt(256));
        mPaint.setColor(color);
        mIndex = index;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        // 圆盘生成基点在屏幕宽度1/4处
        float basePoint =  mContext.getResources()
                .getDisplayMetrics().widthPixels / 4f;
        /*
         * 圆盘索引越大圆盘越宽
         * 圆盘初始宽度
         */
        float left = basePoint - 60 + mIndex * 10;
        float right = basePoint + 60 - mIndex * 10;

        /*
         * 圆盘索引越大圆盘越高
         * 圆盘初始高度
        */
        float top = getTop() + 355 - mIndex * 20;
        float bottom = getTop() + 365 - mIndex * 20;

        // 绘制
        canvas.drawRoundRect(left, top, right, bottom,
                20f, 20f, mPaint);
    }

}

File: app/src/main/java/com/mainxml/hanoitower/DiskView.kt
package com.mainxml.hanoitower

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

/**
 * 汉诺塔 盘子
 * @author zcp 2020/5/23
 */
class DiskView : View {

    private var cornerRadius: Float = dp2px(4).toFloat()
    private var color: Int = Color.BLACK
    private val paint = Paint()

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        context?.apply {
            val attributes = obtainStyledAttributes(attrs, R.styleable.DiskView)
            color = attributes.getColor(R.styleable.DiskView_color, color)
            cornerRadius = attributes.getDimension(R.styleable.DiskView_cornerRadius, cornerRadius)
            attributes.recycle()
        }
        init()
    }

    private fun init() {
        paint.color = color
        paint.apply {
            color = color
            style = Paint.Style.FILL
            isAntiAlias = true
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val defaultWidth = dp2px(56)
        val defaultHeight = dp2px(8)
        val measuredWidth = resolveSize(defaultWidth, widthMeasureSpec)
        val measuredHeight = resolveSize(defaultHeight, heightMeasureSpec)
        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    override fun onDraw(canvas: Canvas?) {
        canvas?.apply {
            val l = paddingLeft.toFloat()
            val t = paddingTop.toFloat()
            val r = paddingRight.toFloat()
            val b = paddingBottom.toFloat()

            val width = width - l - r
            val height = height - t - b

            drawRoundRect(l, t, width, height, cornerRadius, cornerRadius, paint)
        }
    }

    private fun dp2px(dp: Int): Int {
        val scale = context.resources.displayMetrics.density
        return (dp * scale + 0.5f).toInt()
    }
}